From 8a9cb88a7bb57801c1127ccb26e5d5a1f3706ac9 Mon Sep 17 00:00:00 2001
From: Jocelyn Falempe <jfalempe@redhat.com>
Date: Thu, 4 Dec 2025 00:10:47 +0100
Subject: [PATCH] Substitute shell variable from env when starting login

That way, you can specify the right TERM value for agetty
using $$TERM in the systemd unit file.

Signed-off-by: Jocelyn Falempe <jfalempe@redhat.com>
---
 src/pty.c      |  7 +++++-
 src/shl_misc.h | 60 ++++++++++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 66 insertions(+), 1 deletion(-)

diff --git a/src/pty.c b/src/pty.c
index 3039db3f..ee0cf01b 100644
--- a/src/pty.c
+++ b/src/pty.c
@@ -256,6 +256,8 @@ exec_child(const char *term, const char *colorterm, char **argv,
 {
 	char **env;
 	char **def_argv;
+	char **argv_parsed;
+	int ret;
 
 	if (env_reset) {
 		env = malloc(sizeof(char*));
@@ -292,8 +294,11 @@ exec_child(const char *term, const char *colorterm, char **argv,
 	} else {
 		setenv("TERM_SESSION_TYPE", "fb", 1);
 	}
+	ret = shl_replace_array_with_env(&argv_parsed, argv);
+	if (ret)
+		argv_parsed = argv;
 
-	execve(argv[0], argv, environ);
+	execve(argv_parsed[0], argv_parsed, environ);
 
 	log_err("failed to exec child %s: %m", argv[0]);
 
diff --git a/src/shl_misc.h b/src/shl_misc.h
index b994bd3b..72751ea7 100644
--- a/src/shl_misc.h
+++ b/src/shl_misc.h
@@ -413,6 +413,66 @@ static inline int shl_split_command_string(const char *arg, char ***out,
 	return 0;
 }
 
+static inline char *shl_replace_var(char *in)
+{
+	if (!in)
+		return NULL;
+	if (*in != '$')
+		return in;
+	in++;
+	/* if start with $$, remove 1 $ for escaping */
+	if (*in == '$')
+		return in;
+	return getenv(in);
+}
+
+static inline int shl_replace_array_with_env(char ***out, char **argv)
+{
+	char **t, *off;
+	char **nargv;
+	unsigned int size, i, j, len;
+
+	if (!out || !argv)
+		return -EINVAL;
+
+	for (len = 0; argv[len]; ++len)
+		/* empty */ ;
+
+	nargv = malloc(sizeof(*nargv) * (len + 1));
+	if (!nargv)
+		return -EINVAL;
+
+	size = 0;
+	for (i = 0; i < len; ++i) {
+		++size;
+		nargv[i] = shl_replace_var(argv[i]);
+		size += strlen(nargv[i]);
+	}
+	nargv[len] = NULL;
+	++i;
+
+	size += i * sizeof(char*);
+
+	t = malloc(size);
+	if (!t) {
+		free(nargv);
+		return -ENOMEM;
+	}
+	*out = t;
+
+	off = (char*)t + i * sizeof(char*);
+	for (i = 0; i < len; ++i) {
+		*t++ = off;
+		for (j = 0; nargv[i][j]; ++j)
+			*off++ = nargv[i][j];
+		*off++ = 0;
+	}
+	*t = NULL;
+
+	free(nargv);
+	return 0;
+}
+
 static inline int shl_dup_array_size(char ***out, char **argv, size_t len)
 {
 	char **t, *off;
